### PR TITLE
1218 updates for TIM deposit

### PR DESCRIPTION
**Pull request by mwodahl**
Includes updates to SNMP classes relating to the NTCIP1218 spec to allow TIMs to be deposited directly onto 1218-compatible RSUs

Also includes updates to reflect updated 1218 PDU values in SNMPTest and SNMPSessionTest

### CODE REVIEW

File: jpo-ode-core/src/test/java/us/dot/its/jpo/ode/snmp/SNMPTest.java [hunk 1]
```diff
@@ -74,6 +74,6 @@ public void testGettersAndSetters() {
    @Test
    public void testSnmpTimestampFromIso() throws ParseException {
       String snmpTS = SNMP.snmpTimestampFromIso("2017-05-04T21:55:00-05:00");
-      assertEquals("07E105041537", snmpTS);
+      assertEquals("07E1050415370000", snmpTS);
```
Comment (dmccoystephenson) [inline jpo-ode-core/src/test/java/us/dot/its/jpo/ode/snmp/SNMPTest.java:77]: What do these extra 4 zeroes represent?

File: jpo-ode-plugins/src/main/java/us/dot/its/jpo/ode/plugin/SNMP.java [hunk 2]
```diff
@@ -127,6 +127,21 @@ public void setStatus(int status) {
...
+      sb.append(String.format("%02X", zdt.getMonthValue()));
+      sb.append(String.format("%02X", zdt.getDayOfMonth()));
+      sb.append(String.format("%02X", zdt.getHour()));
+      sb.append(String.format("%02X", zdt.getMinute()));
+      sb.append(String.format("%02X", zdt.getSecond()));
+      sb.append(String.format("%02X", zdt.getNano()));
+      return sb.toString();
+   }
+
+   public static String fourOneSnmpTimestampFromIso(String isoTimestamp) throws ParseException {
```
Comment (dmccoystephenson) [inline jpo-ode-plugins/src/main/java/us/dot/its/jpo/ode/plugin/SNMP.java:143]: What's the difference between this `fourOneSnmpTimestampFromIso` method and the `snmpTimestampFromIso` method above?

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/OdeSvcsApplication.java [hunk 3]
```diff
@@ -50,7 +56,13 @@ public static void main(String[] args) throws MalformedObjectNameException, Inte
       mbs.registerMBean(mBean, name);
 
       SecurityProtocols.getInstance().addAuthenticationProtocol(new AuthSHA());
+      SecurityProtocols.getInstance().addAuthenticationProtocol(new AuthHMAC128SHA224());
+      SecurityProtocols.getInstance().addAuthenticationProtocol(new AuthHMAC192SHA256());
+      SecurityProtocols.getInstance().addAuthenticationProtocol(new AuthHMAC256SHA384());
+      SecurityProtocols.getInstance().addAuthenticationProtocol(new AuthHMAC384SHA512());
       SecurityProtocols.getInstance().addAuthenticationProtocol(new AuthMD5());
+      SecurityProtocols.getInstance().addPrivacyProtocol(new PrivAES128());
```
Comment (dmccoystephenson) [inline jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/OdeSvcsApplication.java:64]: Do these protocols automatically get used or do we have to do anything further to utilize them?

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/snmp/SnmpNTCIP1218Protocol.java [hunk 4]
```diff
@@ -86,9 +86,10 @@ public static VariableBinding getVbRsuMsgRepeatPriority(int index) {
     }
 
     public static VariableBinding getVbRsuMsgRepeatOptions(int index) {
+        byte[] val = {(byte) 0xc0};
```
Comment (dmccoystephenson) [inline jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/snmp/SnmpNTCIP1218Protocol.java:89]: This is just defining the byte string in a more correct way, right?

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/snmp/SnmpSession.java [hunk 5]
```diff
@@ -133,9 +134,17 @@ public ResponseEvent set(PDU pdu, Snmp snmpob, UserTarget targetob, Boolean keep
       // Try to send the SNMP request (synchronously)
       ResponseEvent responseEvent = null;
       try {
-         responseEvent = snmpob.set(pdu, targetob);
-         if (!keepOpen) {
-            snmpob.close();
+         byte[] authEngineID = snmpob.discoverAuthoritativeEngineID(targetob.getAddress(), 1000);
```
Comment (dmccoystephenson) [inline jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/snmp/SnmpSession.java:137]: What is this auth engine ID and where does it come from?

**Review by dmccoystephenson — COMMENTED** [review]
This looks good, and the tests pass!

I just have a few clarifying questions.

File: jpo-ode-core/src/test/java/us/dot/its/jpo/ode/snmp/SNMPTest.java (hunk 1 above)
Comment (mwodahl) [reply jpo-ode-core/src/test/java/us/dot/its/jpo/ode/snmp/SNMPTest.java:77]: Those represent seconds + milliseconds

File: jpo-ode-plugins/src/main/java/us/dot/its/jpo/ode/plugin/SNMP.java (hunk 2 above)
Comment (mwodahl) [reply jpo-ode-plugins/src/main/java/us/dot/its/jpo/ode/plugin/SNMP.java:143]: The fourOneSnmpTimestampFromIso method doesn't parse seconds/milliseconds. That extra information isn't required in the FOURDOT1 spec but is required in 1218

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/OdeSvcsApplication.java (hunk 3 above)
Comment (mwodahl) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/OdeSvcsApplication.java:64]: In SnmpSession.java lines 108-109 those protocols are utilized when creating the SNMP user.

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/snmp/SnmpNTCIP1218Protocol.java (hunk 4 above)
Comment (mwodahl) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/snmp/SnmpNTCIP1218Protocol.java:89]: Yep, that just ensures that the value of the byte string is C0

File: jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/snmp/SnmpSession.java (hunk 5 above)
Comment (mwodahl) [reply jpo-ode-svcs/src/main/java/us/dot/its/jpo/ode/snmp/SnmpSession.java:137]: The auth engine ID is the ID of the snmp agent you're sending to (in this case the RSU). This discover method sends a get request to the RSU which then returns its auth engine ID. Without an auth engine ID the snmp4j library will just send an empty PDU.

**Review by dmccoystephenson — APPROVED** [review]
Looks good to me!

**Review by payneBrandon — APPROVED** [review]

**Pull request merged**